### ISSUE TITLE
Fail to create virtualenv for python 2.x.x with ` pyenv virtualenv '.

### ISSUE THREAD

**Issue by hongyi-zhao** [pyenv/pyenv-virtualenv#356]
Hi,

This is Ubuntu 20.04 and git master version of pyenv/pyenv-virtualenv.  Failed to do the following job:
```
$ pyenv shell 2.7.18
$ pyenv virtualenv test
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
ERROR: Could not find an activated virtualenv (required).
```
But for python 3.x.x versions, this problem doesn't appear.

Any hints?

Regards

**Comment by laurent-laporte-pro** [comment]
Actually, you can't create a virtualenv because `pyenv-virtualenv` tries to install the latest version of [virtualenv](https://pypi.org/project/virtualenv/#history), but at the end of 2021 (before Python 2.7 EOL), the last supported version was [16.7.12](https://pypi.org/project/virtualenv/16.7.12/).

So, for this to work, you need to specify version 16.7.12 for virtualenv. There are two ways to do this:

```shell
export VIRTUALENV_VERSION=16.7.12
pyenv virtualenv your_venv-py27
```

The other way to do this is to patch the `pyenv-virtualenv` script as follows:

```diff
@@ -385,14 +385,17 @@
     unset UPGRADE
     VIRTUALENV_OPTIONS[${#VIRTUALENV_OPTIONS[*]}]="--upgrade"
   fi
 fi
 
 if [ -z "${VIRTUALENV_VERSION}" ]; then
   case "${PYENV_VERSION}" in
+  "2.7"* )
+    VIRTUALENV_VERSION="16.7.12"
+    ;;
   "3.0"* )
     NO_ENSUREPIP=1
     ;;
   "3.1"* )
     NO_ENSUREPIP=1
     ;;
   "3.2"* | "stackless-3.2"* )
```

**Comment by hongyi-zhao** [comment]
What about create a PR of this patch?

**Comment by laurent-laporte-pro** [comment]
@hongyi-zhao, Hello, I'm not a bash/fish/zsh developer, I'm a Python developer first of all and so it will be difficult for me to contribute to this project without significant effort: how should I write unit tests to check that the bug fix works well and that there is no regression?

On the other hand, I think the problem is more global: we should be able to make sure that we can install [virtualenv](https://virtualenv.pypa.io/en/latest/index.html) whatever the [version of Python](https://devguide.python.org/versions/#supported-versions) is, including end-of-life versions (which are [not supported anymore](https://devguide.python.org/versions/#unsupported-versions)). We also need to make sure that we use a fresh version of virtualenv so that we can install recent (recent enought) versions of `pip`, `setuptools` and `wheel`.

If we look at the end of life dates of Python versions, we can find a version of virtualenv that fits this date. For example, [virtualenv v16.7.11](https://virtualenv.pypa.io/en/legacy/changes.html#v16-7-11-2021-07-20) was released on 2021-07-20, this is the last version released before the end of life of Python 2.7.18 which was on 2020-01-01.

The table below lists the minor versions of virtualenv v20 that are considered stable:

- v20.16.5 (2022-09-07)
- v20.15.1 (2022-06-28)
- v20.14.1 (2022-04-11)
- v20.13.4 (2022-03-18)
- v20.12.1 (2022-01-01)
- v20.11.2 (2021-12-29)
- v20.10.0 (2021-11-01)
- v20.9.0 (2021-10-23)
- v20.8.1 (2021-09-24)
- v20.7.2 (2021-08-10)
- v20.6.0 (2021-07-14)
- v20.5.0 (2021-07-13)
- v20.4.7 (2021-05-24)
- v20.3.1 (2021-01-13)
- v20.2.2 (2020-12-07)
- v20.1.0 (2020-10-25)
- v20.0.35 (2020-10-15)
- v20.0.0b2 (2020-02-04)
- v20.0.0b1 (2020-01-28)

> It is easy to notice that all new versions have been released after 2020-01-28, so only for Python 3.6
> (the end of life of Python 3.5 took place on 2020-09-30).

The table below lists the legacy versions of [virtualenv](https://virtualenv.pypa.io/en/legacy/changes.html):

- v16.7.11 (2021-07-20)
- v16.6.2 (2019-07-14)
- v16.5.0 (2019-04-24)
- v16.4.3 (2019-03-01)
- v16.3.0 (2019-01-25)
- v16.2.0 (2018-12-31)
- v16.1.0 (2018-10-31)
- v16.0.0 (2018-05-16)
- v15.2.0 (2018-03-21)
- v15.1.0 (2016-11-15)
- v15.0.3 (2016-08-05)
- v14.0.6 (2016-02-07)
- v13.1.2 (2015-08-23)
- v13.0.3 (2015-06-01)
- v12.1.1 (2015-04-07)
- v12.0.7 (2015-02-04)
- v1.11.6 (2014-05-16)
- v1.10.1 (2013-08-07)
- v1.9.1 (2013-03-08)
- v1.8.4 (2012-11-25)
- v1.7.2 (2012-06-22)
- v1.6.4 (2011-07-21)

**Comment by hongyi-zhao** [comment]
You are right. It is not easy to fix this problem systematically and gracefully.

**Comment by native-api** [comment]
> Actually, you can't create a virtualenv because `pyenv-virtualenv` tries to install the latest version of [virtualenv](https://pypi.org/project/virtualenv/#history), but at the end of 2021 (before Python 2.7 EOL), the last supported version was [16.7.12](https://pypi.org/project/virtualenv/16.7.12/).
> 
> So, for this to work, you need to specify version 16.7.12 for virtualenv. There are two ways to do this:
> 

That's strange. This should work, and `pyenv virtualenv test` indeed WFM for 2.7.18 with the latest Pyenv and Pyenv-Virtualenv.

It runs `pip install virtualenv`, and Pip is supposed to pick not just the latest release but the latest release supported by the Python that the running instance of Pip belongs to.

---

First, make sure that `pip` in your environment indeed belongs to the specific Python instance (check the output of `pip --version`)

If it does, please provide a debug trace of the problem:

Run `PYENV_DEBUG=1 PIP_VERBOSE=3 <faulty Pyenv invocation>` and provide the resulting output as a gist.

**Comment by native-api** [comment]
> ERROR: Could not find an activated virtualenv (required).

This error message happens when `PIP_REQUIRE_VIRTUALENV` or `PIP_REQUIRE_VENV` envvars are set, or the corresponding switches are passed to Pip. But `pyenv-virtualenv` unsets those envvars (since a commit from 2014) and you don't supply any other command line options if your report is accurate.

So will need to see the trace to try and find out there they are coming from.
